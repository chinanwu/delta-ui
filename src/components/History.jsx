import PropTypes from 'prop-types';
import React, { useEffect, useRef } from 'react';
import { connect } from 'react-redux';

import getThemeClassname from '../functions/getThemeClassname';

import './History.less';

const makeMysteryStepListItems = num => {
	let all = [];
	for (let i = 0; i < num; i++) {
		all[i] = (
			<li className="History__hintMysteryLi" key={`History__HintMystery--${i}`}>
				<div className="History__hintMystery">?</div>
			</li>
		);
	}
	return all;
};

export const History = ({
	dark,
	to,
	history,
	hintWord,
	hintNumLeft,
	win,
	showHint,
}) => {
	const historyBottomRef = useRef(null);

	useEffect(() => {
		historyBottomRef.current.scrollIntoView({
			behavior: 'smooth',
		});
	}, [history, historyBottomRef]);

	return (
		<div className="History" aria-labelledby="historyLabel">
			<h3 id="historyLabel" className="History__label">
				History
			</h3>
			<ul
				className={
					getThemeClassname('History__list', dark) +
					(win ? ' History__list--hideScroll' : '')
				}
				aria-live="polite"
			>
				{history.map((item, i) => (
					<li key={`historyItem-${i}`}>{item}</li>
				))}
				<li ref={historyBottomRef} />

				{showHint &&
					(hintWord !== to ? (
						<>
							<li className={getThemeClassname('History__hintWord', dark)}>
								{hintWord}
							</li>
							{makeMysteryStepListItems(hintNumLeft - 2)}
							<li>{to}</li>
						</>
					) : (
						<>
							<li className="History__hintWord">{hintWord}</li>
						</>
					))}
			</ul>
		</div>
	);
};

History.propTypes = {
	dark: PropTypes.bool,
	to: PropTypes.string,
	history: PropTypes.arrayOf(PropTypes.string),
	hintWord: PropTypes.string,
	hintNumLeft: PropTypes.number,
	win: PropTypes.bool,
	showHint: PropTypes.bool,
};

export const mapStateToProps = ({ theme: { dark } }) => ({
	dark,
});

export default connect(mapStateToProps)(History);

// TODO:
// - Show hint in history as long as no guess has been entered
