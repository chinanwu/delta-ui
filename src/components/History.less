@import '../style/colours';

.History {
	display: flex;
	flex-direction: column;
	align-items: center;
	text-align: center; // For history label
	height: 100%;
}

.History__label {
	border: 1px solid @green;
	background-color: @green;
	width: 22rem;
	color: black;
}

.History__list {
	border: 1px solid black;
	list-style: none;
	padding: 1rem 0;
	margin: 0;
	overflow: scroll;
	width: 100%;
	height: 100%;
	font-size: 2.4rem;

	//Chrome/Edge/Safari
	&::-webkit-scrollbar,
	&::-webkit-scrollbar-track,
	&::-webkit-scrollbar-corner,
	&::-webkit-scrollbar-thumb {
		background-color: white;
	}

	//Firefox
	scrollbar-color: #b0b0b0 white; /* scroll thumb & track */
}

.History__list--dark {
	border: none;
	background-color: @dark-02;
	color: #fff;

	//Chrome/Edge/Safari
	&::-webkit-scrollbar,
	&::-webkit-scrollbar-track,
	&::-webkit-scrollbar-corner,
	&::-webkit-scrollbar-thumb {
		background-color: @dark-02;
	}

	//Firefox
	scrollbar-color: #323232 @dark-02; /* scroll thumb & track */
}

.History__list--hideScroll {
	overflow: hidden;
}

.History__hintWord {
	color: #757575;
}

.History__hintWord--dark {
	color: #aedb7a;
}

.History__hintMysteryLi {
	display: flex;
	justify-content: center;
}

.History__hintMystery {
	width: 10rem;
	margin-bottom: 0.5rem;
	text-align: center;
	background-color: black;
	color: white;
}

.History__hintMystery--dark {
	background-color: #3c2b3b;
	color: white;
}

@media only screen and (max-width: 550px) {
	.History {
		height: 24.5rem;
	}

	.History__list {
		font-size: 3rem;
		text-align: center;
		padding-left: 0;
		width: 22rem;
	}

	.History__label {
		width: 20rem;
	}
}

@media only screen and (max-width: 320px) and (max-height: 568px) {
	.History {
		height: 19.5rem;
	}

	.History__label {
		padding: 0.5rem 1rem;
	}

	.History__listContainer {
		height: 16rem;
		overflow: scroll;
	}

	.History__list {
		height: 13rem;
	}

	.History__listItem {
		font-size: 2.4rem;
	}
}
