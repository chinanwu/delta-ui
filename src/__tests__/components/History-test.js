import { shallow } from 'enzyme';
import React, { useEffect, useRef } from 'react';

import { History, mapStateToProps } from '../../components/History.jsx';

jest.unmock('../../components/History.jsx');
jest.unmock('../../functions/getThemeClassname');

// mock the HOC
jest.mock('../../components/HOC/withTitle', () => () => Component => props => (
	<Component {...props} />
));

describe('History component', () => {
	describe('rendering', () => {
		it('renders History', () => {
			const ref = { current: { scrollIntoView: jest.fn() } };
			useRef.mockReturnValue(ref);
			const wrapper = shallow(<History history={['test']} />);
			expect(wrapper.getElement()).toMatchSnapshot();
			useEffect.mock.calls[0][0]();
		});

		it('renders hidden scrollbar when win', () => {
			const wrapper = shallow(<History history={['test']} win={true} />);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders hints', () => {
			const wrapper = shallow(
				<History
					history={['test']}
					hintWord="from"
					hintNumLeft={3}
					showHint={true}
				/>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders hint when it is to', () => {
			const wrapper = shallow(
				<History
					history={['test']}
					to="from"
					hintWord="from"
					hintNumLeft={1}
					showHint={true}
				/>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});
	});

	describe('mapStateToProps', () => {
		it('returned mapped properties', () => {
			const theme = { dark: true };
			const state = { theme };
			expect(mapStateToProps(state)).toEqual(theme);
		});
	});
});
