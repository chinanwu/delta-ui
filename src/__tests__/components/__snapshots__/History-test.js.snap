// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`History component rendering renders History 1`] = `
<div
  aria-labelledby="historyLabel"
  className="History"
>
  <h3
    className="History__label"
    id="historyLabel"
  >
    History
  </h3>
  <ul
    aria-live="polite"
    className="History__list"
  >
    <li>
      test
    </li>
    <li />
  </ul>
</div>
`;

exports[`History component rendering renders hidden scrollbar when win 1`] = `
<div
  aria-labelledby="historyLabel"
  className="History"
>
  <h3
    className="History__label"
    id="historyLabel"
  >
    History
  </h3>
  <ul
    aria-live="polite"
    className="History__list History__list--hideScroll"
  >
    <li>
      test
    </li>
    <li />
  </ul>
</div>
`;

exports[`History component rendering renders hint when it is to 1`] = `
<div
  aria-labelledby="historyLabel"
  className="History"
>
  <h3
    className="History__label"
    id="historyLabel"
  >
    History
  </h3>
  <ul
    aria-live="polite"
    className="History__list"
  >
    <li>
      test
    </li>
    <li />
    <React.Fragment>
      <li
        className="History__hintWord"
      >
        from
      </li>
    </React.Fragment>
  </ul>
</div>
`;

exports[`History component rendering renders hints 1`] = `
<div
  aria-labelledby="historyLabel"
  className="History"
>
  <h3
    className="History__label"
    id="historyLabel"
  >
    History
  </h3>
  <ul
    aria-live="polite"
    className="History__list"
  >
    <li>
      test
    </li>
    <li />
    <React.Fragment>
      <li
        className="History__hintWord"
      >
        from
      </li>
      <li
        className="History__hintMysteryLi"
      >
        <div
          className="History__hintMystery"
        >
          ?
        </div>
      </li>
      <li />
    </React.Fragment>
  </ul>
</div>
`;
